### DISCUSSION TITLE
Assembler rp2040 instructions

### DISCUSSION THREAD

**Discussion by digilabpg123** [micropython/micropython#10401] [RP2040 / Pico]
hi, I'm writing a micropython code with Assembler instructions, I saw that the MOVW MOVT and MOVWT instructions are not accepted, if it is possible to have some info or the right syntax. thanks to whoever can help me

**Comment by robert-hh** [comment]
There is some documentation here: https://docs.micropython.org/en/latest/pyboard/tutorial/assembler.html?highlight=assembler
and here: https://docs.micropython.org/en/latest/reference/asm_thumb2_index.html#asm-thumb2-index

**Reply by digilabpg123** [reply]
thanks for the answer but I think they only work on STM32 CPUs I tried on RP2040 they don't seem to work, I wanted to know if anyone had succeeded

**Comment by th3w** [comment]
A list of ARM thumb instructions supported by the RP2040 can be found in the RP2040 datasheet starting on page 69.
https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf

**Reply by digilabpg123** [reply]
thanks for the answer, as I thought those three instructions do not seem to exist

**Comment by russhughes** [comment]
samneggs has several micropython/rp2040 assembly code examples that I found useful.
https://gist.github.com/samneggs 
https://www.youtube.com/@samneggs1

**Reply by samneggs** [reply]
Thanks for the shoutout, Russ.

**Comment by peterhinch** [comment]
I have a number of libraries written for STM (Arm Thumb V7) which I adapted for the Pico (V6). In my experience the differences are minor, aside from the missing floating point instructions. There is a restriction on the stack push and pop instructions - IIRC they won't handle registers above R7 but aside from that my code worked.

**Reply by samneggs** [reply]
Speaking of high registers, I’d like to see some assembly instructions added. My focus is the RP2040 and I’ve resorted to using op-codes to access MOV, ADD and CMP on the high registers using the data directive. There’s also a useful byte reverse instruction for swapping low/high byte of a register I use for RGB565. 
I wonder if these could be added?

**Reply by peterhinch** [reply]
I jumped through similar hoops a few years ago to cope with the (then) lack of floating point instructions on STM. I bit the bullet and submitted a PR adding the current FP subset. I suggest you do likewise. For it to be accepted I advise including tests and doc updates.

As a general point I wrote a [driver for SSD1351](https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/ssd1351/ssd1351.py) with a key `_lcopy` routine written in Assembler. I then wrote a [portable version](https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/ssd1351/ssd1351_generic.py) with the routine coded using Viper. To my great surprise the Viper code is almost as fast. Viper is amazingly good at integer and bit manipulation coding.

**Reply by samneggs** [reply]
That’s some nice code, Peter. I also wrote some assembly routines to convert one bit-plane and 4 bit-planes to RGB565. I fully agree using viper routines with integer math to approach C level performance. I think your work inspired me to go in that direction some time ago. 
As far as MicroPython changes, that looks like too much work for what is a minor inconvenience for me. It also may be cool to be one of the very few to be using inline assembly’s high registers.

**Reply by scruss** [reply]
Incidentally, a user on the Raspberry Pi forum (who can't/won't raise issues here) has noted some bugs in the rp2 inline assembler, particularly with the interpretation of variables as registers:
[pio program stopping after 6 entries - Raspberry Pi Forums](https://forums.raspberrypi.com/viewtopic.php?p=2076248#p2075847)

**Reply by samneggs** [reply]
I believe that post about the bugs was referring to the pio assembler, not the ARM thumb assembler as discussed above. Both are inline assemblers I suppose.
